src/rng/backends/curand/mrg32k3a.cpp

```diff
@@ -0,0 +1,956 @@
+                auto acc = r.get_access<cl::sycl::access::mode::read_write>(cgh);
+                cgh.codeplay_host_task([=](cl::sycl::interop_handle ih) {
+                    auto r_ptr =
+                        reinterpret_cast<float*>(ih.get_native_mem<cl::sycl::backend::cuda>(acc));
+                    curandStatus_t status;
+                    CURAND_CALL(curandGenerateNormal, status, engine_, r_ptr, n, distr.mean(),
+                                distr.stddev());
+                });
+            })
+            .wait_and_throw();
+    }
+
+    virtual void generate(const oneapi::mkl::rng::gaussian<
+                              double, oneapi::mkl::rng::gaussian_method::box_muller2>& distr,
+                          std::int64_t n, cl::sycl::buffer<double, 1>& r) override {
+        queue_
+            .submit([&](cl::sycl::handler& cgh) {
+                auto acc = r.get_access<cl::sycl::access::mode::read_write>(cgh);
+                cgh.codeplay_host_task([=](cl::sycl::interop_handle ih) {
+                    auto r_ptr =
+                        reinterpret_cast<double*>(ih.get_native_mem<cl::sycl::backend::cuda>(acc));
+                    curandStatus_t status;
+                    CURAND_CALL(curandGenerateNormalDouble, status, engine_, r_ptr, n, distr.mean(),
+                                distr.stddev());
+                });
+            })
+            .wait_and_throw();
+    }
+
+    virtual void generate(
+        const oneapi::mkl::rng::gaussian<float, oneapi::mkl::rng::gaussian_method::icdf>& distr,
+        std::int64_t n, cl::sycl::buffer<float, 1>& r) override {
+        queue_
+            .submit([&](cl::sycl::handler& cgh) {
+                auto acc = r.get_access<cl::sycl::access::mode::read_write>(cgh);
+                cgh.codeplay_host_task([=](cl::sycl::interop_handle ih) {
+                    auto r_ptr =
+                        reinterpret_cast<float*>(ih.get_native_mem<cl::sycl::backend::cuda>(acc));
+                    curandStatus_t status;
+                    CURAND_CALL(curandGenerateNormal, status, engine_, r_ptr, n, distr.mean(),
+                                distr.stddev());
+                });
+            })
+            .wait_and_throw();
+    }
+
+    virtual void generate(
+        const oneapi::mkl::rng::gaussian<double, oneapi::mkl::rng::gaussian_method::icdf>& distr,
+        std::int64_t n, cl::sycl::buffer<double, 1>& r) override {
+        queue_
+            .submit([&](cl::sycl::handler& cgh) {
+                auto acc = r.get_access<cl::sycl::access::mode::read_write>(cgh);
+                cgh.codeplay_host_task([=](cl::sycl::interop_handle ih) {
+                    auto r_ptr =
+                        reinterpret_cast<double*>(ih.get_native_mem<cl::sycl::backend::cuda>(acc));
+                    curandStatus_t status;
+                    CURAND_CALL(curandGenerateNormalDouble, status, engine_, r_ptr, n, distr.mean(),
```

Comment:
I am not sure if[ icdf method ](https://spec.oneapi.com/versions/latest/elements/oneMKL/source/domains/rng/distributions-template-parameter-mkl-rng-method-values.html#onemkl-rng-distributions-template-parameter-mkl-rng-method-values)should call the same curandGenerateNormalDouble as it should use different mathematical methods for gaussian transform. (Unfortunately, I haven't found in cuRAND documentation, what method is used). I would suggest adding implementation only in the method, which is used as by_default(box_muller2).

Reply:
Indeed, I noticed this as well; the inverse cumulative distribution function method is only used for quasirandom number generators in cuRAND, e.g.:
https://docs.nvidia.com/cuda/curand/group__HOST.html#group__HOST_1g56ff2b3cf7e28849f73a1e22022bcbfd

I was a bit surprised the tests passed, but since they did, I kept them. I will unimplement functions using ICDF methods.

Reply:
The test is passed, as it checks only statistical properties, which are ok for both methods. But it should be fine also if it throws an unimplemented exception - the test would be just skipped.